[Merged by Bors] - feat: basic linear algebra lemmas about matrices

ocfnash opened the pull request:
These came up in the proof that the Geck Construction yields semisimple Lie algebras. I am splitting them into their own PR to simplify review.


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat: basic linear algebra lemmas about matrices

These came up in the proof that the Geck Construction yields semisimple Lie algebras. I am splitting them into their own PR to simplify review.

ocfnash committed:
Relax `NonUnitalNonAssocCommRing` to `NonUnitalNonAssocCommSemiring` as per CR suggestion

ocfnash commented on Mathlib/Data/Matrix/Mul.lean:
```diff
@@ -411,6 +411,11 @@ theorem diagonal_mul_diagonal' [Fintype n] [DecidableEq n] (d₁ d₂ : n → α
     diagonal d₁ * diagonal d₂ = diagonal fun i => d₁ i * d₂ i :=
   diagonal_mul_diagonal _ _
 
+theorem commute_diagonal {α : Type*} [NonUnitalNonAssocCommRing α]
```
Good point, done!

j-loreaux commented on Mathlib/LinearAlgebra/Eigenspace/Matrix.lean:
```diff
@@ -69,6 +69,50 @@ lemma hasEigenvalue_toLin'_diagonal_iff [NoZeroDivisors R] (d : n → R) {μ : R
+    simp [Pi.single_apply]
+
+@[simp]
+lemma iSup_maxGenEigenspace_diagonal_eq_top :
+    ⨆ μ, maxGenEigenspace (diagonal d).toLin' μ = ⊤ := by
```
Likewise, you can prove that for `i : n`, `eigenspace (diagonal d) μ = span (Pi.basisFun R n '' (d ⁻¹' μ))`. Or else you can just use the result in the other comment and [hasEigenvector_toLin'_diagonal](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Eigenspace/Matrix.html#hasEigenvector_toLin'_diagonal) to prove that every element of `Pi.basisFun` is in the LHS.

ocfnash replied:
As above, I've incorporated this suggestion.

ocfnash commented on Mathlib/Data/Matrix/Block.lean:
```diff
@@ -216,6 +216,14 @@ theorem fromBlocks_multiply [Fintype l] [Fintype m] [NonUnitalNonAssocSemiring 
   rcases i with ⟨⟩ <;> rcases j with ⟨⟩ <;> simp only [fromBlocks, mul_apply, of_apply,
       Sum.elim_inr, Fintype.sum_sum_type, Sum.elim_inl, add_apply]
 
+theorem fromBlocks_pow [Semiring α] [Fintype n] [Fintype m] [DecidableEq n] [DecidableEq m]
+    (A : Matrix n n α) (D : Matrix m m α) (k : ℕ) :
+    (fromBlocks A 0 0 D) ^ k = fromBlocks (A ^ k) 0 0 (D ^ k) := by
```
Thanks, I prefer `fromBlocks_diagonal_pow` for the reason you have outlined.

ocfnash committed:
CR suggestion: establish `maxGenEigenspace` is just `eigenspace` and use instead

ocfnash committed:
Cater for `simp` set change

Also minor golfs

ocfnash committed:
Fix broken proof

The issues here was a non-terminal `simp` (exposed by adding
`Basis.sum_repr` to the `simp` set).

ocfnash committed:
whitespace fix

ocfnash committed:
Nitpick style tweaks

j-loreaux reviewed (commented):
Ah, this is much nicer! Thanks a lot!

bors merge
